Fixes for #1208 and partial #1196

giig1967g opened the pull request:
My first pull request.
Please forgive me if I do something wrong.

- Fixed #1208.

- #1196: Added some format commands for 0/1:
#C: 1/0 = Open/Close
#U: 1/0 = Up/Down

giig1967g committed:
Fixes for #1208 and partial #1196

Fixed #1208
Added dìformat commands:
#C: 1/0 = Open/Close
#U: 1/0 = Up/Down

giig1967g commented:
Hi all,
I can't understand what I did wrong.
I would like to contribute. Can someone explain what went wrong?
thanks

Oxyandy commented:
In the logs

```
/home/travis/build/letscontrolit/ESPEasy/src/ESPEasy.ino:398:16: error: variable or field 'WiFiEvent' declared void
PluginCall(PLUGIN_FIFTY_PER_SECOND, 0, dummyString);
^
/home/travis/build/letscontrolit/ESPEasy/src/ESPEasy.ino:398:16: error: 'system_event_id_t' was not declared in this scope
/home/travis/build/letscontrolit/ESPEasy/src/ESPEasy.ino:398:41: error: 'system_event_info_t' was not declared in this scope
PluginCall(PLUGIN_FIFTY_PER_SECOND, 0, dummyString);
^
```

giig1967g commented:
But I didn't change that file.
I changed only misc.ino
I can't understand.

TD-er commented:
There is a problem with the PlatformIO builder since the last update of @mvdbro 
I've been trying on my system also to get it fixed, but there is something very strange happening to the parsing of `#ifdef ESP32`.

It is very well explained here: https://github.com/platformio/platformio-core/issues/801#issuecomment-252447178

However, I do not yet have a solution and until then all builds will fail.

But I'm working on it.

giig1967g commented:
ok. :)
So as soon as it is fixed it will be automatically rebuilt or I have to resubmit it?

TD-er commented:
I will then trigger a rebuild, or merge it via my own computer instead of direct merge

TD-er commented:
Just created a PR for the build issues. https://github.com/letscontrolit/ESPEasy/pull/1212
Just have to wait and see if it will also build on the Travis servers.

TD-er commented on src/Misc.ino:
```diff
@@ -1561,19 +1561,37 @@ String parseTemplate(String &tmpString, byte lineSize)
 
                         int oidx;
-                        if ((oidx = valueFormat.indexOf('O')) >= 0) // Output
+                        if (value.toFloat() == 0 || value.toFloat() == 1) //if val != 0 or != 1 ignore the format setting
```
toFloat() is a rather expensive operation. Better store it in a variable and compare that variable instead of doing it twice. (next line does it again)
And also comparing a float to an int (or another float) is tricky. For example: 0.00000001 != 0
The comment describes the opposite of the code.

When the user wants such a discrete format, doesn't it make more sense to output < 0,5 als the "0" equivalent and >0.5 as the "1" ?

giig1967g replied:
regarding toFloat(), you are right. Will change that.
The reason why I did the check on 0 or 1 only is because the #O format is valid only if value is 1 or 0, so for GPIO, not for any float value.

TD-er replied:
But you're "casting" it to float. So the result may be a float and comparing floats to int is always giving issues.
Like `if (a == 0)` will return false if `a` equals 0.0000001.
So you can "round" it to an int (`int(a + 0.5)`) before comparing, or state `if (a < 0.5) { return "0";}`

giig1967g replied:
The toFloat() operation was already in the original code.
And it was already doing the casting.
What then if i use toInt() instead of toFloat()?

giig1967g commented:
Any news on the merge?

TD-er commented:
@giig1967g I did write a few questions about your PR as inline remarks.
So I'm actually waiting for a reply on those questions.

giig1967g commented:
Hi, I am a bit lost...
Where are your comments? :(

TD-er commented:
Apparently I had to press "submit review".

giig1967g committed:
modified float() call

giig1967g commented:
Sent a new PR with a change in the toFloat call

TD-er commented:
But still comparing the float to an int.
You really should not do that using "=="
Only compare floats using "<", ">".
If you want to compare a float (or double) you should check if it is within a range. like ` if(a > (b - (b* eps)) && a < (b+(b*eps)))`
Where eps is some small value where 1.0 + eps != 1.0

So please keep that in mind, because comparing floating point values is always a good receipt for funky errors.

uzi18 commented:
Why not check only if value!=0.0 then show on else show off,  why so strict
0,1?

czw., 5.04.2018, 00:59 użytkownik Gijs Noorlander <notifications@github.com>
napisał:

> But still comparing the float to an int.
> You really should not do that using "=="
> Only compare floats using "<", ">".
> If you want to compare a float (or double) you should check if it is
> within a range. like if(a > (b - (b* eps)) && a < (b+(b*eps)))
> Where eps is some small value where 1.0 + eps != 1.0
>
> So please keep that in mind, because comparing floating point values is
> always a good receipt for funky errors.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/letscontrolit/ESPEasy/pull/1210#issuecomment-378771215>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHOUxlk2Ku907psGpBb_7sXuKTFuzHPks5tlVBRgaJpZM4TCCCL>
> .
>

TD-er commented:
@uzi18
My suggestion was to check for <0.5 or something similar.
The rest is just for information on why you must never use == to compare floats.

TD-er commented:
N.B. != Is the same operator as == only negated.

uzi18 commented:
@TD-er I prefer more C-like check.
0=false, everything other true.

giig1967g commented:
Here is my proposal:
                        int val = value.toFloat() < 0.5 ? 0 : 1;

TD-er commented:
@uzi18 Me too, but my point was that you simply may not check floating point values with `==` of `!=`.
There are very specific use cases for those checks, but for all other cases where you use `==` on floats, it is simply a source of very funky errors at runtime.
That's why I'm so keen on those, when I see them in source code.

@giig1967g `const int val = value.toFloat() < 0.5 ? 0 : 1;` is fine.

giig1967g commented:
changed (and simplified) the whole logic to:
Wil try it tonight.


```
                        const int val = value.toFloat() < 0.5 ? 0 : 1;

                        if (valueFormat == "O")
                          value = val == 0 ? "OFF" : " ON";
                        else if (valueFormat == "O!")
                          value = val == 0 ? " ON" : "OFF";
                        else if (valueFormat == "C")
                          value = val == 0 ? "CLOSE" : " OPEN";
                        else if (valueFormat == "C!")
                          value = val == 0 ? " OPEN" : "CLOSE";
                        else if (valueFormat == "U")
                          value = val == 0 ? "DOWN" : "  UP";
                        else if (valueFormat == "U!")
                          value = val == 0 ? "  UP" : "DOWN";
                        else if (valueFormat == "R") {
                          int filler = lineSize - newString.length() - value.length() - tmpString.length() ;
                          for (byte f = 0; f < filler; f++)
                            newString += " ";
                          }

```

TD-er commented:
Currently the valueFormat contains all from the hastag to the end of the string
https://github.com/giig1967g/ESPEasy/blob/765c56ccefe429dceb8421374e4f3916953d6083/src/Misc.ino#L1525-L1529

This can also be part of this string: `[task#sensor]`, so please also test with sensor values named "U", "C", etc.

giig1967g commented:
I am checking... but apparently it's not: 
Consider this:
```
tmpString=[task#value#F]
deviceName=task
valueName=value
valueFormat=F
```

uzi18 commented:
Try also #!OR formating it should work

uzi18 commented:
Please add also
```
                        else if (valueFormat == "Y")
                          value = val == 0 ? " NO" : "YES";
```

giig1967g commented:
this should work. 
Compiles fine, but didn't test it yet.

```
const int val = value.toFloat() < 0.5 ? 0 : 1; //to be used for GPIO status (0 or 1)
const int inverted = valueFormat.indexOf('!') >= 0 ? 1 : 0;

if (valueFormat.indexOf('O') >= 0)
  value = val == inverted ? "OFF" : " ON"; //(equivalent to XOR operator)
else if (valueFormat.indexOf('C') >= 0)
  value = val == inverted ? "CLOSE" : " OPEN";
else if (valueFormat.indexOf('U') >= 0)
  value = val == inverted ? "DOWN" : "  UP";
else if (valueFormat.indexOf('Y') >= 0)
  value = val == inverted ? " NO" : "YES";

if (valueFormat.indexOf('R') >= 0) {
  int filler = lineSize - newString.length() - value.length() - tmpString.length() ;
  for (byte f = 0; f < filler; f++)
    newString += " ";
}

```

giig1967g commented:
the latest code accepts the '!' in any position and 'R' could be combined with the other format settings.

giig1967g committed:
new #formats

Hi, here the new formats:
#O: ON/OFF
#C: Open/Close
#U: UP/Down
#Y: YES/NO
#D: converto to nearest integer
#F: convert to floor
#E: convert to ceiling

TD-er reviewed (commented):
This will always return the "1" version (or the "0" when inverted) when number of decimals> 0.

giig1967g committed:
added more formats

#X
#y
#Z

giig1967g commented:
Hi @TD-er ,

> This will always return the "1" version (or the "0" when inverted) when number of decimals> 0.

This is correct for the formats O,C,U,Y,y,X,Z as they should be used ONLY for GPIO status, so 0 or 1
`value = val == inverted ? " NO" : "YES";`
For formats D,E,F the value returned is the float value formatted.
`value = (int)roundf(valFloat);`

giig1967g commented:
I also added a check before the format transformation in order to avoid the loop if there is no format set.

I think this can be merged. What do you think?

giig1967g commented:
for the Wiki:

**Format Transformation:**

Feature to transform Values from numeric to strings or to format numeric values.
Syntax:
[task#value#format]

Where #format is:
Format of PIN states of GPIO (0 or 1)
#O: 1/0 -> ON/OFF
#C: 1/0 ->  OPEN/CLOSE
#U: 1/0 ->  UP/DOWN
#Y: 1/0 ->  YES/NO
#I: 1/0 ->  IN/OUT
#y: 1/0 ->  Y/N
#X: 1/0 ->  X/O
#Z: 1/0 ->  1/0
Adding ! to any of the above formatting inverts the logic.
So for example:
#O! 1/0 -> OFF/ON
#Z! 1/0 -> 0/1
#Y! 1/0 -> NO/YES


Format of any number value (floats):
#D: convert to nearest integer 
#E: convert to ceiling 
#F: convert to floor 

Justification formats:
#R: justifies the latest value to the right according to the number of characters exported by the LCD (tested on LCD). Not tested on other displays.

Grovkillen commented:
Will add once merged. :+1:

Grovkillen commented:
Should we have "IN/OUT" as well?

giig1967g committed:
Added #I

#I: IN/OUT

giig1967g commented:
Added #I 
Updated wiki text

I suggest to freeze it for now.
We might add other formats later.

TD-er commented:
```C++
        const int val = value == "0" ? 0 : 1; //to be used for GPIO status (0 or 1)
                          const float valFloat = value.toFloat();
                          const int inverted = valueFormat.indexOf('!') >= 0 ? 1 : 0;

                          if (valueFormat.indexOf('O') >= 0)
                            value = val == inverted ? "OFF" : " ON"; //(equivalent to XOR operator)

```

I was just looking at the code.
Is the position of the matching character always the same?
Then you can simply do a switch statement and thus save a lot of `indexOf()` calls.

And the invert-check logic can also be done as first logic.


For example: (little bit of pseudo code, not sure about position N)

```C++
                          const float valFloat = value.toFloat();
                          int val = valFloat < 0.5 ? 0 : 1; //to be used for GPIO status (0 or 1)
                          if (valueFormat.indexOf('!') >= 0) {
                            val = (val == 0 ? 1 : 0);
                          }
                          // Must check length of valueFormat, just to be sure, for index N of the character
                          if (valueFormat.length() >= N) {
                          switch (valueFormat[N]) {
                             case 'O':
                                value = val ? "OFF" : " ON"; //(equivalent to XOR operator)
                                break;
                             case .....

```

giig1967g commented:
That was the original coding... but that could have broken backwards compatibilty, as the format could be:
#O!R or #RO! or #!RO so there is no fixed position.

Also consider that this is a secondary function that will be executed only if there is a format definition.
Otherwise it will be skipped:
`if (valueFormat.length() > 0) //do the checks only if a Format is defined to optimize loop
`

giig1967g commented:
any hope of seeing this merged with the master?

TD-er commented:
Yes there is, but I've been too busy fixing the wifi issues and last week there has been some personal issues taking a lot of time and energy. (related to government and our houses)

giig1967g commented:
Thanks a lot.
